cupy/lib/stride_tricks.py

```diff
@@ -37,3 +38,110 @@ def as_strided(x, shape=None, strides=None):
+    window_shape : int or tuple of int
+        Size of window over each axis that takes part in the sliding window.
+        If `axis` is not present, must have same length as the number of input
+        array dimensions. Single integers `i` are treated as if they were the
+        tuple `(i,)`.
+    axis : int or tuple of int, optional
+        Axis or axes along which the sliding window is applied.
+        By default, the sliding window is applied to all axes and
+        `window_shape[i]` will refer to axis `i` of `x`.
+        If `axis` is given as a `tuple of int`, `window_shape[i]` will refer to
+        the axis `axis[i]` of `x`.
+        Single integers `i` are treated as if they were the tuple `(i,)`.
+    subok : bool, optional
+        If True, sub-classes will be passed-through, otherwise the returned
+        array will be forced to be a base-class array (default).
+    writeable : bool, optional
+        When true, allow writing to the returned view. The default is false,
+        as this should be used with caution: the returned view contains the
+        same memory location multiple times, so writing to one location will
+        cause others to change.
+
+    Returns
+    -------
+    view : ndarray
+        Sliding window view of the array. The sliding window dimensions are
+        inserted at the end, and the original dimensions are trimmed as
+        required by the size of the sliding window.
+        That is, ``view.shape = x_shape_trimmed + window_shape``, where
+        ``x_shape_trimmed`` is ``x.shape`` with every entry reduced by one less
+        than the corresponding window size.
+
+
+    See also
+    --------
+    numpy.lib.stride_tricks.as_strided
+
+    Notes
+    --------
+    This function is adapted from numpy.lib.stride_tricks.as_strided.
+
+    Examples
+    --------
+    >>> x = _cupy.arange(6)
+    >>> x.shape
+    (6,)
+    >>> v = sliding_window_view(x, 3)
+    >>> v.shape
+    (4, 3)
+    >>> v
+    array([[0, 1, 2],
+           [1, 2, 3],
+           [2, 3, 4],
+           [3, 4, 5]])
+
+    """
+    window_shape = (tuple(window_shape)
+                    if np.iterable(window_shape)
+                    else (window_shape,))
+
+    # first convert input to array, possibly keeping subclass
+    x = _cupy.array(x, copy=False, subok=subok)
+
+    window_shape_array = _cupy.array(window_shape)
+    if _cupy.any(window_shape_array < 0):
+        raise ValueError('`window_shape` cannot contain negative values')
+
+    if axis is None:
+        axis = tuple(range(x.ndim))
+        if len(window_shape) != len(axis):
+            raise ValueError(f'Since axis is `None`, must provide '
+                             f'window_shape for all dimensions of `x`; '
+                             f'got {len(window_shape)} window_shape elements '
+                             f'and `x.ndim` is {x.ndim}.')
+    else:
+        axis = np.core.numeric.normalize_axis_tuple(axis, x.ndim)
```

Comment:
`np.core.numeric.normalize_axis_tuple` is not NumPy's public API. Please use `cupy._core.internal._normalize_axis_indices` instead.

Reply:
Yes, we initially implemented `cupy._core.internal._normalize_axis_indices` instead of the numpy version. However, the behaviors of these two are slightly different. Unfortunately `cupy._core.internal._normalize_axis_indices` do not accept axis as a list while the numpy equivalent handle the list and this result in an error when using cupy-xarray. 
Please see this discussion in this PR : https://github.com/cupy/cupy/pull/7575#discussion_r1204707652

Here is the error when using the cupy `_nomralize_axis_indices` : 
```
File cupy/core/internal.pyx:403, in cupy.core.internal._normalize_axis_indices()

File cupy/core/internal.pyx:424, in cupy.core.internal._normalize_axis_indices()

TypeError: 'list' object cannot be interpreted as an integer
```

Reply:
`axis` should be a tuple of int, not a list of int.
https://github.com/cupy/cupy/blob/25c485071ab0a33cdc223c793aa049ab4d84c097/cupy/_core/internal.pyx#L447-L449

Reply:
I have updated xarray to pass axis as a tuple not a list. [Xarray/PR#7983](https://github.com/pydata/xarray/pull/7938)

I also made the corresponding changes here to use `cupy._core.internal._normalize_axis_indices` as you asked. 
This is fully addressed now.